### PR TITLE
Shopping-Cart: Show Taxes (Placeholder edition)

### PR DESCRIPTION
**Pull request by deBhal**
#### Changes proposed in this Pull Request

This PR injects placeholder values into the shopping cart as it goes to/from the server, and renders those values in the shopping cart components. It is stacked on #27922.

These changes are stirring up quite a few style issues in pre-existing code, but I think I'll push those out to a separate PR

#### Testing instructions

Enable `show-tax` in your config or pass the `?flags=show-tax` query param to one of the pages that show the shopping cart:

checkout_ _julesauspremiumtest_ _wordpress_com_and___a8c_wp-calypso_client_my-sites_checkout_cart_style_scss_ _wp-calypso

domains_ _julesauspremiumtest_ _wordpress_com

and the checkout mobile view:

checkout_ _julesauspremiumtest_ _wordpress_com

We don't have responses from the server yet, so for now we need to check manually that we're sending the right values. Look for `POST` api calls to `/me/shopping-cart`, and check that they're sending the values you'd expect in `tax.location`:

checkout_ _julesauspremiumtest_ _wordpress_com

You should also be able to see the pay button be disabled while we're processing updates (it's _much_ faster against the production servers, but my sandbox is nice and slow and obvious):

checkout pending updates

Also check the zip validation. A postal code is always required, but in the US it must be a 5 digit number:

checkout postal code validation

#### TODO in a followup PR:

- [ ] make the payment methods names constants
- [ ] factor out shopping cart payment accessors and similar logic that operates on the store but doesn't depend on it

### CODE REVIEW

**Comment by deBhal** [comment]
Well, that's unfortunate:
checkout_ _julesauspremiumtest_ _wordpress_com

It was mostly like that when I got here, but definitely needs fixing now
checkout_ _julesauspremiumtest_ _wordpress_com

**Comment by deBhal** [comment]
@blowery: this code is still rough, but I wanted to ask if this is the best approach to the flux store - I've just pushed a "proper" flux approach, but the values I want are already in the redux store.

The values I want are already in the redux store (or will be after #27922). Would it make reduxifying easier later if I used `reduxGetState()` from `redux-bridge` to pull those values out of the redux store and inject them into the cart in response to the flux action instead?

Is there some other alternative that would be better?

**Comment by sixhours** [comment]
This looks solid from a design standpoint! I had a few notes, no blockers.

The 30% width on the divider in the cart sidebar feels out of place; I think it's because it's both shorter and a different color than other dividers in that section. Extending it to full width matches the other dividers and I think it looks more balanced. What do you think?

screen shot 2018-11-07 at 2 05 09 pm

You already noted the issue with the help menu on mobile; adding some padding to the bottom of the cart is a hacky fix, but it's definitely more usable:

screen shot 2018-11-07 at 2 53 52 pm

Or a bottom margin, for comparison:

screen shot 2018-11-07 at 2 53 25 pm

For the price buttons, I'd consider adding a bit more breathing room for readability. For example, I've added 2px top margin to the price, and 1px left margin to the superscript here. It's super subtle, but I think it helps everything look less cramped:

screen shot 2018-11-07 at 2 04 16 pm

**Comment by deBhal** [comment]
> The 30% width on the divider in the cart sidebar feels out of place; I think it's because it's both shorter and a different color than other dividers in that section. Extending it to full width matches the other dividers and I think it looks more balanced. What do you think?

@sixhours: I agree the 30% could use improving, but I don't think the 100% is quite right either - The other full-width bars are dividing separate parts of the cart from each other, but the total is very closely related to the subtotals, so we want something to separate the total from the subtotals but still keep them grouped together separate from the domain or the subscription.

Do you know what I mean? I was thinking about the underline in a maths problem with that first design, but maybe we could do something with colour? Or match the size of the line to the text to make that association stronger? I'm just throwing ideas out here, hoping you've got a better one :)

**Comment by deBhal** [comment]
> For the price buttons, I'd consider adding a bit more breathing room for readability. For example, I've added 2px top margin to the price, and 1px left margin to the superscript here. It's super subtle, but I think it helps everything look less cramped:

Yeah, it is subtle, but you can see the improvement. Here's the before:
price space before
and after:
price space after

>  adding some padding to the bottom of the cart is a hacky fix, but it's definitely more usable

Simple but effective. I went with the padding over the margin, but it's a soft preference:

checkout_ _julesauspremiumtest_ _wordpress_com_and_activity_monitor__all_processes__and_automattic____notes__-_vim

**Comment by sixhours** [comment]
> Do you know what I mean? I was thinking about the underline in a maths problem with that first design, but maybe we could do something with colour?

My next suggestion was going to be to lighten it up a bit. :D Not as light as the upper divider, since that seems to fade away, but one degree less (so if you're using `$gray`, switch to `$gray-lighten-10` or `$gray-lighten-20`). Something like this:

screen shot 2018-11-08 at 10 32 42 am

>That's 65px, but I wonder if I should add even more, to make space for the icon in its entirety. ( ... 

I probably wouldn't, in order to try to balance readability with the amount of empty space at the bottom. The icon can overlap the sidebar, just not the text.

> Tangentially: Is it just me, or is the question mark not quite centered in that button?)

Something's definitely up with that icon. I wonder if it's a browser thing, since it looks OK to me in Chrome?

File: client/lib/cart/store/index.js [hunk 1]
```diff
@@ -151,6 +165,47 @@ CartStore.dispatchToken = Dispatcher.register( payload => {
...
+			break;
+
+		case TRANSACTION_PAYMENT_SET:
+			{
+				// set both
+				let postalCode, countryCode;
+
+				const paymentMethod = get( action, [ 'payment', 'paymentMethod' ] );
+				switch ( paymentMethod ) {
+					case 'WPCOM_Billing_MoneyPress_Stored':
```
Comment (yoavf) [inline client/lib/cart/store/index.js]: Not a blocker, but worth making the payment methods names constants at some point

Comment (deBhal) [reply client/lib/cart/store/index.js]: Yeah, that's a good idea. It fits nicely into the other refactor I want to do too.

The problem I ran into was that I need a place to put stuff - we can't have circular dependencies from the stores requiring each other.

The shopping cart has `cart-values` for all the the operations on the store, so I created `transactions-values` and started putting stuff there and the diff started growing really fast, but that's the perfect place to put these constants.

**Comment by deBhal** [comment]
>  Not as light as the upper divider, since that seems to fade away, but one degree less

Heh, I went one degree lighter (in https://github.com/Automattic/wp-calypso/pull/28212/commits/03b5d4062bd2f9a56a708211742f413b37d261ce), turns out that _is_ as light as the upper divider, and looks like this:

checkout_ _julesauspremiumtest_ _wordpress_com_and_cross_reference___trunk_bin_tests_domains_api_opensrs-api-tests_php
domains_ _julesauspremiumtest_ _wordpress_com
checkout_ _julesauspremiumtest_ _wordpress_com

That's what in there right now, and I think I'm going to ship that behind the feature flag, but I'm keen to iterate a bit more if you are.

I was wondering about just making it shorter. Reducing the width behaves differently with the mobile version, so that didn't really work, but I like the result of using a table to make the line match the text, it just needs a bit more fiddling and a lot more cross-browser testing, so I've broken it out into https://github.com/Automattic/wp-calypso/pull/28453 with screenshots. Do you like that better?

File: client/lib/cart/store/index.js [hunk 2]
```diff
@@ -151,6 +165,47 @@ CartStore.dispatchToken = Dispatcher.register( payload => {
...
+						postalCode = extractStoredCardMetaValue( action, 'card_zip' );
+						countryCode = extractStoredCardMetaValue( action, 'country_code' );
+						break;
+					case 'WPCOM_Billing_MoneyPress_Paygate':
+						const paymentDetails = get( action, 'payment.newCardDetails', {} );
+						postalCode = paymentDetails[ 'postal-code' ];
+						countryCode = paymentDetails.country;
+						break;
+					default:
+						throw new Error( 'Unrecognized payment method', action );
```
Comment (yoavf) [inline client/lib/cart/store/index.js]: Sorry I missed this - there are other payment methods (like `WPCOM_Billing_WPCOM`) and this caused those payments to fail. (Reverted)

Comment (deBhal) [reply client/lib/cart/store/index.js]: Thanks - I'll fix add appropriate tests and try again

**Pull request merged**